jwcrypto/pkcs11.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (C) 2018 JWCrypto Project Contributors - see LICENSE file
+
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+try:
+    from gnutls.library.constants import GNUTLS_DIG_SHA1, GNUTLS_DIG_SHA256, \
+        GNUTLS_DIG_SHA384, GNUTLS_DIG_SHA512
+    from gnutls.crypto import PrivateKey
+    from gnutls.callbacks import gnutls_set_pin_for_keyurl, \
+        gnutls_remove_pin_for_keyurl
+    found_gnutls = True
+except ImportError:
+    found_gnutls = False
+
+
+def _hashfn_to_gnutls(hashfn):
+    if isinstance(hashfn, hashes.SHA1):
+        return GNUTLS_DIG_SHA1
+    if isinstance(hashfn, hashes.SHA256):
+        return GNUTLS_DIG_SHA256
+    if isinstance(hashfn, hashes.SHA384):
+        return GNUTLS_DIG_SHA384
+    if isinstance(hashfn, hashes.SHA512):
+        return GNUTLS_DIG_SHA512
+
+
+def bytes_to_long(barray):
+    i = 0
+    res = 0
+    func = ord
+    if type(barray).__name__ == 'bytes':
+        # on python3 we get bytes...
+        func = int
+    while i < len(barray):
+        res = res * 256 + func(barray[i])
+        i = i + 1
+    return res
+
+
+class PKCS11Key(object):
+
+    def __init__(self):
+        self.gnutls_key = None
+
+    @classmethod
+    def import_from_pkcs11_url(cls, url, pin=None):
+        if not found_gnutls:
+            raise ImportError('python-gnutls not found')
+        if pin:
+            gnutls_set_pin_for_keyurl(url, pin)
+        pk = cls()
+        try:
+            pk.gnutls_key = PrivateKey.import_url(url)
+        finally:
+            if pin:
+                gnutls_remove_pin_for_keyurl(url)
+        return pk
+
+    def get_url(self):
+        return self.gnutls_key.get_url()
+
+    def get_public_key(self):
+        return PKCS11PublicKey(self.gnutls_key.get_public_key())
+
+    def sign(self, payload, _padfn, hashfn):
+        return self.gnutls_key.sign_data(_hashfn_to_gnutls(hashfn),
+                                         0, payload)
+
+    def decrypt(self, ct, _padfn):
+        return self.gnutls_key.decrypt_data(0, ct)
+
+
+class PKCS11PublicKey(object):
+
+    def __init__(self, gnutls_pubkey):
+        self.gnutls_pubkey = gnutls_pubkey
+
+    def public_numbers(self):
+        n, e = self.gnutls_pubkey.export_rsa_raw()
+        return rsa.RSAPublicNumbers(bytes_to_long(e), bytes_to_long(n))
```

Comment:
it would be really nice to be able to use int.from_bytes(barray, byteorder='little') instead of this function, unfortunately from_bytes is not available in python2.
So please instead try to use int(hexlify(n.reverse()), 16) and remove bytes_to_long entirely.

Reply:
With hexlify I get this problem on python2: `AttributeError: 'str' object has no attribute 'reverse'`
python3 sees bytes: `AttributeError: 'bytes' object has no attribute 'reverse'`

Reply:
`int(hexlify(e), 16)` works